replay: C3/C3X hardware decoder

jakethesnake420 opened the pull request:
On the comma device, kill openpilot managed procs in tmux with ctrl+c, then start the ui again with `./selfdrive/ui/ui`. 
then run:
`COMMA_CACHE=/data/cache ./tools/replay/replay --demo --ecam --dcam`

jakethesnake420 commented:
@adeebshihadeh this is the full working implementation. There are still some issues though.

jakethesnake420 commented:
@deanlee when running replay on the device with --demo --no-vipc, i see the model path and lane lines stuttering roughly every 5 seconds.

adeebshihadeh reviewed (changes requested):
Very cool! Did a quick pass and left some comments to start.

Do we really need the rotator? Decoder won't output NV12?

(To clarify, I'd like to merge the rotator support anyway. It's nice to have support for all the HW blocks in openpilot anyway.)

jakethesnake420 commented:
I am basing this off a working example code https://github.com/[jakethesnake420/v4l2-decode](https://github.com/jakethesnake420/v4l2-decode). With that said,  I don't have a complete understanding of why certain things are done the way they are.

---

> Do we really need the rotator? Decoder won't output NV12?

The venus driver throws an error when trying to set the capture buffer type to NV12.

```
msm_vidc:   fw: 
               <VFW_E:HostDr:265d:92a139e6:0> vDec_LoadResources(1228): Linear format not supported for DPB!
```

```
comma@comma-fa9d355d:/$ strings firmware/image/venus.* | grep Linear
%s(%d): Linear format not supported for DPB!
%s(%d): Linear format not supported for DPB!

```
Which tells me that the decoded picture buffer needs to be UBWC. The DPB format seems to get overridden when I set the capture plane format. I have tried a few things to get it working but the only way I could see is to use the rotator. 

---

The stuff in decoder.h is copied from kernel headers to work around build issues. I think this can be cleaned up I just have not spent the time yet to solve the issue.

---

The change to msgq_repo is to add a modified version of the VisionBuf::allocate that maps the memory differently. 

new function allocate_no_cache:
```
void VisionBuf::allocate_no_cache(size_t length) {
ion_alloc.len = length;
ION_IOC_ALLOC
ION_IOC_MAP
ION_IOC_FREE
mmap
```

versus original allocate:
```
void VisionBuf::allocate(size_t length) {
ion_alloc.len = length + PADDING_CL + sizeof(uint64_t);
ION_IOC_ALLOC
ION_IOC_SHARE
mmap

calls ION_IOC_FREE later
```

---

>rotate can't handle this?

Another pain point. I couldn't get the driver to accept a custom fmt_cap.fmt.pix.bytesperline as its calculated by the driver and its based on the resolution and alignment. I tried to format the output to a size that the driver would accept and then scale it up/down but I faced additional issues. I am not an expert on troubleshooting v4l2 or video formats so its difficult for me to tell what is wrong when an image is deformed a certain way.

jakethesnake420 commented on tools/replay/qcom/sde_rotator.cc:
```diff
@@ -0,0 +1,171 @@
+#define V4L2_PIX_FMT_NV12_UBWC v4l2_fourcc('Q', '1', '2', '8')
+#endif
+
+static void checked_ioctl(int fd, unsigned long request, void *argp) {
```
What if safe_ioctl accepts a message and an error callback? It is used in spi.cc so it would be like this:

`util::safe_ioctl(spi_fd, SPI_IOC_WR_MODE, &spi_mode, "failed setting SPI mode", fail);`

adeebshihadeh commented:
@jakethesnake420 can you rebase after the ioctl improvements and do a bit of cleanup? then I can see if I see can simplify this a bit before merging

adeebshihadeh commented:
trigger-jenkins

Commit:
bump msgq

Commit:
add third_party/linux/include/media/msm_vidc.h

Commit:
add sde_rotator hw interface

Commit:
add msm_vidc hw decoder interface

Commit:
update SConscript to build qcom decoder and rotator

Commit:
use qcom decoder in replay framereader

jakethesnake420 commented:
> can you rebase after the ioctl improvements and do a bit of cleanup? then I can see if I see can simplify this a bit before merging

Rebased, cleaned and further simplified.

jakethesnake420 commented:
The last major thing to simplify is the visionbuf stuff. Maybe allocate() can be refactored slightly to work.

adeebshihadeh committed:
Merge branch 'master' of ssh://github.com/commaai/openpilot into qcom-replay

jakethesnake420 commented:
Ok I will take another swing at these issues.

jakethesnake420 committed:
decode directly to NV12 with the correct stride without using the hw rotator

Commit:
bump msgq back to master

jakethesnake420 commented on tools/replay/qcom/decoder.cc:
```diff
@@ -259,7 +262,7 @@ bool MsmVidc::setDBP() {
   struct v4l2_ext_controls controls = {0};
   control[0].id           = V4L2_CID_MPEG_VIDC_VIDEO_STREAM_OUTPUT_MODE;
-  control[0].value        = 0; // V4L2_CID_MPEG_VIDC_VIDEO_STREAM_OUTPUT_PRIMARY
+  control[0].value        = 1; // V4L2_CID_MPEG_VIDC_VIDEO_STREAM_OUTPUT_SECONDARY
```
This was the key difference to get the decoder to decode directly to linear NV12 (non-UBWC)

Commit:
don't compile rotator

adeebshihadeh committed:
Merge branch 'master' of ssh://github.com/commaai/openpilot into qcom-replay

adeebshihadeh committed:
cleanup

adeebshihadeh committed:
Merge branch 'master' of ssh://github.com/commaai/openpilot into qcom-replay

adeebshihadeh committed:
works now but much to simplify

adeebshihadeh committed:
rm signals

jakethesnake420 commented:
It should be possible to use the ion buffers created by vipc, eliminating the copy buffer and buffer creation code. But it may not be simple.

adeebshihadeh committed:
rm header

adeebshihadeh commented:
Thanks for the PR!

jakethesnake420 commented:
Your welcome!
